File: src/system/generate_system_files.py
#!/usr/bin/env python3

import jinja2
import json
import pathlib
import sys

THIS_DIR = pathlib.Path(__file__).parent.absolute()
SCHEMA_PATH =  THIS_DIR / "schema.json"
TEMPLATE_PATH = THIS_DIR / "templates"
TEMPLATE_NAME = sys.argv[1]
OUTPUT_PATH_FORMAT = sys.argv[2];

class SqParam:
    def __init__(self, index, schema):
        self.name = schema["name"]
        self.index = index
        self.base_type = schema["type"]
        self.required = schema["required"]
        if self.required:
            self.type = f"const {self.base_type}&"
        else:
            self.type = f"const {self.base_type}*"

class SqField:
    def __init__(self, name, schema):
        self.name = name
        self.return_type = schema["return"]
        self.params = [SqParam(i, s) for i, s in enumerate(schema["params"])]

class SqType:
    def __init__(self, name, schema):
        self.name = name
        self.fields = [SqField(n, s) for n, s in schema.items()]


def get_schema():
    with SCHEMA_PATH.open(mode="rb") as s:
        return json.load(s)

def generate_system_files():

    sq_types = [SqType(n, s) for n, s in get_schema().items()]

    jinja_env = jinja2.Environment(
        loader=jinja2.FileSystemLoader(TEMPLATE_PATH),
        autoescape=False,
        undefined=jinja2.StrictUndefined
    )
    template = jinja_env.get_template(f"{TEMPLATE_NAME}.jinja");

    for sq_type in sq_types:
        path = pathlib.Path(OUTPUT_PATH_FORMAT % sq_type.name)
        path.parent.mkdir(parents=True, exist_ok=True);
        with open(path, "wt") as f:
            f.write(template.render(sq_type=sq_type))

generate_system_files()

File: src/system/generate_system_files.lua
#!/bin/env lua
-- -----------------------------------------------------------------------------
-- Copyright 2021 Jonathan Haigh
-- SPDX-License-Identifier: MIT
-- -----------------------------------------------------------------------------

schema_path = arg[1]
template_path = arg[2]
output_path_format = arg[3]

function sq_param(param_index, param_schema)
    local ptr_or_ref = param_schema["required"] and "&" or "*"
    local param_type = "const " .. param_schema["type"] .. ptr_or_ref
    return {
        name = param_schema["name"],
        index = param_index,
        base_type = param_schema["type"],
        required = param_schema["required"],
        type = param_type,
    }
end

function sq_field(field_name, field_schema)
    local field_params = {}
    for index, param_schema in pairs(field_schema["params"]) do
        table.insert(field_params, sq_param(index - 1, param_schema))
    end
    return {
        name = field_name,
        return_type = field_schema["return"],
        params = field_params,
    }
end

function sq_type(type_name, type_schema)
    local field_tables = {}
    for field_name, field_schema in pairs(type_schema) do
        table.insert(field_tables, sq_field(field_name, field_schema))
    end
    return {
        name = type_name,
        fields = field_tables,
    }
end

function get_schema()
    local lunajson = require("lunajson")
    local f = assert(io.open(schema_path, "rb"))
    local schema = lunajson.decode(f:read("*all"))
    f:close()
    return schema
end

function compile_template(liluat, path)
    local f = assert(io.open(path, "rb"))
    local template = liluat.compile(f:read("*all"))
    f:close()
    return template
end

function generate_system_files()
    local liluat = require("liluat")
    local template = compile_template(liluat, template_path)
    for type_name, type_schema in pairs(get_schema()) do
        local rendered = liluat.render(template, sq_type(type_name, type_schema))
        local output_path = string.format(output_path_format, type_name)
        local f = assert(io.open(output_path, "wb"))
        f:write(rendered)
        f:close()
    end
end

generate_system_files()
